core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java

```diff
@@ -198,15 +199,17 @@ private static int indexOfUnsafeOctet(String s, int start) {
         return -1;
     }
 
-    // Do not inline `index` or remove `index < 0` check.
-    // They are specifically written to tell JVM to optimize away the array boundary check.
-    @SuppressWarnings({ "ConstantConditions", "UnnecessaryLocalVariable" })
     private static boolean isUnsafeOctet(char c) {
-        final int index = c;
-        if (index < 0 || index >= SAFE_OCTETS.length) {
-            return true;
+        if (c >= SAFE_OCTETS.length) {
+            return false;
```

Comment:
The original logic is a bug?
It returns `true` if `index >= SAFE_OCTETS.length`. 🤔

Reply:
> This PR also fixes the off-by-one error 😱

Is this? 🤣

Reply:
Nope. We had to return `true` here :scream: